File: src/superenhancers/super_enhancer.R
#!/usr/bin/env Rscript

suppressMessages(library(data.table))
suppressMessages(library(docstring))


 calculate_cutoff <- function(vector) {
    #' Calculate superenhancer signal cut-off by sliding a diagonal line until the tangent is found
    #' 
    #' @param vector matrix Background corrected stitched enhancer signal values matrix 
    #' 
    #' @return List of superenhancer signal cut-off values


	#Get the slope of the line to slide
 	vector <- vector[order(vector)]
	slope <- (max(vector) - min(vector)) / length(vector)

	#Find the x-axis point where a line passing through that point has the minimum number of points below it (ie. tangent)
	xPt <- floor(optimize(numPts_below_line, lower=1, upper=length(vector), vector=vector, slope=slope)$minimum)
	y_cutoff <- vector[xPt]

	return(list(absolute=y_cutoff, overMedian=y_cutoff/median(vector), overMean=y_cutoff/mean(vector)))
}


numPts_below_line <- function(vector, slope, x) {
    #' Calculate linear equation of diagonal at point [x,y] and the number of signal values below the diagonal
    #' 
    #' @param vector matrix Background corrected stitched enhancer signal values matrix 
    #' @param slope double Diagonal slope
    #' @param x double X point at which to calculate the linear equation
    #' 
    #' @return Number of signal values below the diagonal at point [x,y]
    
    #Calculate lienar equation (y = ax+b)
	y <- vector[x]
	b <- y-(slope*x)

    #Calculate number of signal values below diagonal
	xPts <- 1:length(vector)
	return(sum(vector <= (xPts*slope+b)))
}

File: src/superenhancers/super_enhancer.py
#!/usr/bin/env python

import math
import numpy as np

from scipy.optimize import minimize_scalar


def calculate_cutoff(
    vector: np.ndarray
) -> np.float:
    """Calculate superenhancer signal density cutoff value

    Args:
        vector (np.ndarray): Array of (control corrected) stitched enhancer loci density signals

    Returns:
        np.float: Density signal cutoff value to delineate superenhancers from normal enhaners 
    """

    #Get the slope of the line to slide
    vector.sort()
    slope = (max(vector) - min(vector)) / len(vector)

    #Minimising the (control corrected) stitched enhancer loci density signal function (aka finding the tangent of the function)
    x_min = math.floor(minimize_scalar(numPts_below_line, bounds=(1, len(vector)), args=(vector, slope), method="bounded")["x"])
    y_cutoff = vector[x_min]

    return y_cutoff


def numPts_below_line(
    x: float,
    vector: np.ndarray,
    slope: np.float64
) -> int:
    """Stitched enhancer loci density signal function to minimise

    Args:
        x (float): Signal density value to perform the calculation at
        vector (np.ndarray): Signal density values vector
        slope (np.float64): Slope coefficient

    Returns:
        int: Number of signal density values equal to or smaller than a line going through point x 
    """

    #Calculate line equation at point x   (y = ax+b)
    y = vector[int(x)-1]
    b = y-(slope*x)

    #Calculate y values of points on the line and the number of points that are larger than those of the vector
    xPts = np.array(range(1, len(vector)+1))
    return sum(vector <= (xPts*slope+b)) 